### DISCUSSION TITLE
ESM only packages: Remix should automatically bundle all dependencies of the mentioned packages

### DISCUSSION THREAD

**Discussion by HuyAms** [remix-run/remix#2594] [Proposals]
### What version of Remix are you using?

1.3.2

### Steps to Reproduce

Hi, I try to add `react-markdown` to Remix and get this error `Error: require() of ES Module`. To fix the issue, I follow this doc: https://remix.run/docs/en/v1/api/conventions#serverdependenciestobundle and add `react-markdown`there but it doesn't help.

```
serverDependenciesToBundle: [
    /^react-markdown.*/,
],
```

https://stackblitz.com/edit/node-u2wf1e?file=app%2Froutes%2Findex.tsx

### Expected Behavior

I expect to fix the issue by just adding `/^react-markdown$/`to the serverDependenciesToBundle and Remix will automatically bundle all the related dependencies

### Actual Behavior

It turns out that I have to add all the dependencies to the serverDependenciesToBundle to make it work

```
serverDependenciesToBundle: [
        /^rehype.*/,
        /^remark.*/,
        /^unified.*/,
        /^unist.*/,
        /^hast.*/,
        /^bail.*/,
        /^trough.*/,
        /^mdast.*/,
        /^micromark.*/,
        /^decode.*/,
        /^character.*/,
        /^property.*/,
        /^space.*/,
        /^comma.*/,
        /^react-markdown$/,
        /^vfile.*/,
        /^ccount*/,
        /^markdown-table*/,
    ]
   ```

**Comment by jacob-ebey** [comment]
We have a new utility exported from `@remix-run/dev` called `getDependenciesToBundle` that can be used to get the package, and all it's dependencies. It will get more resilient to misconfigured package.json's in the next release as well. You can use it like this:

```js
const { getDependenciesToBundle } = require("@remix-run/dev");

module.exports = {
  serverDependenciesToBundle: [
    ...getDependenciesToBundle(
      "react-markdown",
      "any-other-top-level-packages-you-import-and-need-to-bundle"
    )
  ],
};
```

I personally would like to avoid doing anything automatic here as it may lead to a case where a package references one of those deps and isn't bundled, leading to two copies, the bundled one, and a dynamically imported one from node_modules being used at runtime.

**Reply by Bushuo** [reply]
Hi, i just found this discussion, but it seems like `getDependenciesToBundle` does not function correctly.
As soon as I started using it I get this waring in the console `path/to/my-project/node_modules/package.json does not exist`
It seems that the function does not traverse the dependencies correctly.

Inspecting in the debugger yields only one entry inside `aggregatedDeps`
![image](https://user-images.githubusercontent.com/9990578/165401956-7ba0830d-c58d-48f8-becd-f500b224eb58.png)

Furthermore `package.json` that don't provide a `main` will produce an error.
See this repo that is a sub-dependency of `react-markdown` [https://github.com/micromark/micromark/tree/main/packages/micromark-util-symbol](https://github.com/micromark/micromark/tree/main/packages/micromark-util-symbol)

Also packages that somehow have `@types/` dependecies can not be resolved.

I would be happy to provide a PR for all of this, as i worked it out locally, if someone could help me with the test cases :)

**Reply by alindsay55661** [reply]
@Bushuo Are you willing to share your fixes on another branch while you await help with tests? I could really use this.

**Reply by Bushuo** [reply]
@alindsay55661 This is the branch I will try to merge after I wrote the tests.
https://github.com/Bushuo/remix/tree/fix-getDependeciesToBundle

**Reply by Bushuo** [reply]
This is the PR that shows one of those bugs #3106
I hope somebody can help me with the tests and then we merge my other branch into this PR branch 🥳

**Reply by alindsay55661** [reply]
Helpful, thank you!

**Reply by ZipBrandon** [reply]
I am currently in the same boat with `react-markdown`, so thank you as well @Bushuo!

**Reply by kwiat1990** [reply]
I use Remix `1.5.1` and I can't get `react-markdown` and some 'rehype' plugins to work. I have tried both ways of including ESM modules but in each case Remix start complaining about other submodules imported, I guess, in these modules, which I have previously defined.

**Reply by ZipBrandon** [reply]
@kwiat1990 - I am using all of those deps. Here is an excerpt of my `remix.config.js` file (I have a few extra ESM deps too but this shouldn't impact you).

```js
module.exports = {
   // ....
  serverDependenciesToBundle: [
    /^remark.*/,
    /^rehype.*/,
    /^unified.*/,
    /^micromark.*/,
    /^mdast.*/,
    /^react-querybuilder.*/,
    /^react-day-picker.*/,
    /^react-dnd-html5-backend.*/,
    /^dnd-core.*/,
    /^@react-dnd.*/,
    /^@udecode\/plate-ui-dnd.*/,
    /^@udecode\/plate.*/,
    /^@udecode\/plate-serializer-md.*/,
    /^decode-named-character-reference.*/,
    /^character-entities.*/,
    /^unist-util-stringify-position.*/,
    /^character-reference-invalid.*/,
    /^is-decimal.*/,
    /^is-hexadecimal.*/,
    /^is-alphanumerical.*/,
    /^copy-to-clipboard.*/,
    /^unist-.*/,
    /^hast-util-parse-selector.*/,
    /^property-information.*/,
    /^hast-util-whitespace.*/,
    /^space-separated-tokens.*/,
    /^comma-separated-tokens.*/,
    /^zen-observable-ts.*/,
    /^ts-invariant.*/,
    /^zen-observable-ts.*/,
    /^is-alphabetical.*/,
    /^@sindresorhus\/slugify.*/,
    /^react-markdown.*/,
    /^micromark-util-symbol.*/,
    /^zwitch.*/,
    /^bail.*/,
    /^trough.*/,
    /^fault.*/,
    /^ccount.*/,
    /^react-dnd.*/,
    /^parse-entities.*/,
    /^stringify-entities.*/,
    /^hastscript.*/,
    /^ts-invariant.*/,
    /^unist-util-map.*/,
    /^markdown-table.*/,
    /^longest-streak.*/,
    /^vfile.*/,
    /^vfile-message.*/,
    /^@apollo\/client.*/,
    /^@sindresorhus\/transliterate.*/,
  ],
};
```

**Reply by kwiat1990** [reply]
@ZipBrandon, that was exactly my problem: I needed to add some more packages to the list. This seemed at first a bit odd to me that also not directly imported packages should be on the list but I understand it’s the way it work right know. With that I could solve all problems and get it to work.

**Reply by dbstraight** [reply]
I had to also add `/^trim-lines.*/` and then this "worked" for me. "Worked" is a strange word to use though for having to manually list all dependencies recursively in order to use modern ESM-only packages.

Any update from the team on making `getDependenciesToBundle` work?

**Reply by kiliman** [reply]
In the meantime I have a script you can use:
https://github.com/kiliman/rmx-cli#-get-esm-packages

**Reply by dbstraight** [reply]
Tried it. Does not find all required dependencies for `react-markdown`.

Remix 1.7, React 18.2, React-Markdown 8.0.3

Generates this list:

```
"bail",
"character-entities",
"comma-separated-tokens",
"decode-named-character-reference",
"hast-util-whitespace",
"mdast-util-definitions",
"mdast-util-from-markdown",
"mdast-util-to-hast",
"mdast-util-to-string",
"micromark",
"micromark-core-commonmark",
"micromark-factory-destination",
"micromark-factory-label",
"micromark-factory-space",
"micromark-factory-title",
"micromark-factory-whitespace",
"micromark-util-character",
"micromark-util-chunked",
"micromark-util-classify-character",
"micromark-util-combine-extensions",
"micromark-util-decode-numeric-character-reference",
"micromark-util-decode-string",
"micromark-util-encode",
"micromark-util-html-tag-name",
"micromark-util-normalize-identifier",
"micromark-util-resolve-all",
"micromark-util-sanitize-uri",
"micromark-util-subtokenize",
"micromark-util-symbol",
"micromark-util-types",
"property-information",
"react-markdown",
"remark-parse",
"remark-rehype",
"space-separated-tokens",
"trough",
"unified",
"unist-builder",
"unist-util-generated",
"unist-util-is",
"unist-util-position",
"unist-util-stringify-position",
"unist-util-visit",
"unist-util-visit-parents",
"vfile",
"vfile-message"
```

And when you try to start the server, get this error:

```
micromark is possibly an ESM only package and should be bundled with "serverDependenciesToBundle in remix.config.js.
trim-lines is possibly an ESM only package and should be bundled with "serverDependenciesToBundle in remix.config.js.
...
Error [ERR_REQUIRE_ESM]: require() of ES Module...
```

**Reply by Unintendedz** [reply]
I'm stuck at micromark too ..

**Reply by rishi-raj-jain** [reply]
Yeah, I bundled using @vercel/nft so two suggestions:
- Remix config maybe support async calls -> using nodeFileTrace is an async operation
- micromark even though included, comes out with the warning as:
```
micromark is possibly an ESM only package and should be bundled with "serverDependenciesToBundle" in remix.config.js.
```

**Reply by kirandash** [reply]
after following @kiliman's suggestion on running the cli `npx rmx-cli get-esm-pack`.  I was getting `micromark is possibly an ESM only package` error. 

I had to add `/^micromark.*/,` to fix the error.

**Reply by kiliman** [reply]
Yes, my script simply follows the package dependencies looking for ESM modules. Packages like `micromark` which supports dynamic plugins tend to be an issue. So instead of explicit package names, you may need to use the regex like above.

**Comment by ryanflorence** [comment]
You can tell Remix to bundle everything, I tend to do this 😅

```js
module.exports = {
  serverDependenciesToBundle: [/.*/],
}
```

**Reply by 0xi4o** [reply]
Well this worked for me. Using this at least till we get official support.

**Reply by jdaly13** [reply]
Is there a downside to this?  Any performance issues? npm dependency hell it seems like everyone is experiencing.  Stupid wagmi got rid of common JS but remix keeps telling me I need to include these bundles even though I have included them so obviously they are not tracking dependencies properly and this is a bug.

**Reply by markhughes** [reply]
Easiest way to get unblocked until someone from the dev team can give some pointers on how to go about debugging this, cheers!

**Reply by ariofrio** [reply]
This seems to break some other packages, in turn:

```
▲ [WARNING] "./xhr-sync-worker.js" should be marked as external for use with "require.resolve" [require-resolve-not-external]

    node_modules/jsdom/lib/jsdom/living/xhr/XMLHttpRequest-impl.js:31:57:
      31 │ ... require.resolve ? require.resolve("./xhr-sync-worker.js") : null;
         ╵                                       ~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] Could not resolve "./lib-cov/fluent-ffmpeg"

    node_modules/fluent-ffmpeg/index.js:1:56:
      1 │ ...FMPEG_COV ? require('./lib-cov/fluent-ffmpeg') : require('./lib/...
        ╵                        ~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] Could not resolve "canvas"

    node_modules/jsdom/lib/jsdom/utils.js:160:25:
      160 │   const Canvas = require("canvas");
          ╵                          ~~~~~~~~

  You can mark the path "canvas" as external to exclude it from the bundle, which will remove this error. You can also surround this "require" call with a try/catch block to handle this failure at run-time instead of bundle-time.

```

**Reply by ariofrio** [reply]
To fix that, use the following regex-foo to exclude certain packages but bundle all others:

```
serverDependenciesToBundle: [/^(?!.*(jsdom|fluent-ffmpeg)).*$/]
```

This even fixes [this issue I was having with "serverDependenciesToBundle" seeming to ignore "mdast-util-to-markdown"](https://github.com/remix-run/remix/issues/109#issuecomment-1583083330)!

**Comment by sebruiz** [comment]
Our team was able to use a library that depends on `react-markdown` called `@uiw/react-md-editor` by taking advantage of the `<ClientOnly>` wrapper and the `.client.tsx` file extension. Maybe this helps someone else.

Relevant PR:
https://github.com/MetricsDAO/xyz/pull/466

**Reply by AlemTuzlak** [reply]
You're a life saver! I was literally trying to make it work with the same package and after adding `.client.tsx` it worked! thank you!
